src/coins.cpp

```diff
@@ -171,7 +171,14 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn
                 // Create the coin in the parent cache, move the data up
                 // and mark it as dirty.
                 CCoinsCacheEntry& entry = cacheCoins[it->first];
-                entry.coin = std::move(it->second.coin);
+                if (erase) {
+                    // The `move` call here is purely an optimization; we rely on the
+                    // `mapCoins.erase` call in the `for` expression to actually remove
+                    // the entry from the child map.
+                    entry.coin = std::move(it->second.coin);
```

Comment:
Thinking about this more, is it not more than a pure optimization? If we copy instead of move, then we are increasing the memory footprint of our dbcache. For a large batch write won't all dirty entries exist in memory twice, potentially increasing the memory requirements of the node temporarily until the parent cache writes to disk?

Reply:
Hmm ok I guess I'm missing how we're erasing during the iteration with `erase=false`. I tested a reindex-chainstate with max dbcache and instead did a `Sync` instead of `Flush` on shutdown, and it didn't increase memory usage at all. LGTM.

Reply:
@andrewtoth you don't see a big memory spike because in practice this implementation of `BatchWrite` is only called for small sets of coins [during `ConnectTip()`](https://github.com/jamesob/bitcoin/blob/8d5994ef2235c6706440bc18968006cd04c22fb4/src/validation.cpp#L2600-L2613) to make UTXO set changes atomic (i.e. a CCoinsViewCache on top of another CCoinsViewCache); calling `BatchWrite` for really sizable sets of coins happens on `CCoinsViewDB` (leveldb) objects (i.e. writing down from a CCoinsViewCache -> CCoinsViewDB). CCoinsViewDB doesn't have an in-memory cache (at least that we maintain) so there is no sizable duplication.

But once again, thanks for raising the point and testing.